### PR TITLE
:sparkles: :recycle: Switch to Express and add React

### PR DESCRIPTION
**Pull request by kronicker**
- Switched backend to Express.js
- Made client basic rendering in React
- Added `webpack` for bundling
- Added `bootstrap` with bootswatch theme

p.s.: Dont forget to npm install!

@DorianSyruss, feel free to fork after this PR is accepted and take over styling and adding missing functionality

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      filename: path.join(__dirname, 'dist/index.html'),
+      template: path.join(__dirname, 'client/index.html'),
+      inject: true
+    }),
+    new ExtractTextPlugin('style.css')
+  ]
+};
+
+module.exports = env => {
+  let config = Object.assign({}, baseConfig);
```
Comment (vladimyr) [inline webpack.config.js]: Technically you don't need to copy config object unless you are going to modify it? 🤔

File: webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  if (env === 'dev') {
+    config.devtool = 'eval';
+    config.devServer = {
+      contentBase: path.join(__dirname, "dist"),
+      noInfo: true,
+      proxy: {
+        "/api": "http://localhost:3030"
+      }
+    }
+  }
```
Comment (vladimyr) [inline webpack.config.js]: Either flip this `if` so you end up with early return or extract this chunk to helper function.

File: webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      query: {
+        limit: 10000,
+        name: '[name].[ext]?[hash]'
+      }
+    }, {
+      test: /\.(woff2?|svg)$/,
+      loader: 'url-loader',
+      query: {
+        limit: 10000
+      }
```
Comment (vladimyr) [inline webpack.config.js]: ⚠️  This is solely an artistic observation
If you have only one property inline it in order to save vertical space (less scrolling = 😀 ):
```js
query: { limit: 10000 }
```

File: server/index.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+const port = process.env.PORT || 3030;
+
+
+calendar.get('/api/calendar', getCalendar);
+app.use('/api/calendar', getCalendar),
+app.use('/', express.static(path.join(__dirname, '../dist')))
+
+
+app.listen(port);
+console.log(`Server listening on port ${port}`);
```
Comment (vladimyr) [inline server/index.js]: Hm, `#listen(port, [calback])` takes second param which is callback firing as soon as server actually starts listening so put this log inside it.

File: server/index.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+const calendar = require('express').Router();
+const path = require('path');
+const app = express();
+const getCalendar = require('./calendar');
+const port = process.env.PORT || 3030;
+
+
+calendar.get('/api/calendar', getCalendar);
+app.use('/api/calendar', getCalendar),
+app.use('/', express.static(path.join(__dirname, '../dist')))
```
Comment (vladimyr) [inline server/index.js]: Lines 11 and 12 mean exactly the same with following difference:
- 11 adds handler to `calendar` router
- 12 adds handler to global router ie `app`

That's redundant and besides that you are not even using `calendar` anywhere.

File: webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+module.exports = env => {
+  let config = Object.assign({}, baseConfig);
+  if (env === 'dev') {
+    config.devtool = 'eval';
+    config.devServer = {
+      contentBase: path.join(__dirname, "dist"),
+      noInfo: true,
+      proxy: {
+        "/api": "http://localhost:3030"
```
Comment (vladimyr) [inline webpack.config.js]: Try to somehow share port value between express and webpack dev server, possibly through custom field in `package.json`?

File: package.json [hunk 7]
```diff
@@ -13,8 +16,34 @@
...
+    "jquery": "^3.2.1",
+    "node-sass": "^4.5.2",
+    "react": "^15.5.4",
+    "react-dom": "^15.5.4",
+    "resolve-url-loader": "^2.0.2",
+    "sass-loader": "^6.0.4",
+    "style-loader": "^0.17.0",
+    "url-loader": "^0.5.8",
+    "webpack": "^2.5.1",
+    "webpack-dev-server": "^2.4.5"
```
Comment (vladimyr) [inline package.json:50]: Hmm, as far as I understand things following packages should be moved to dependencies map:
- axios
- date-fns
- jquery
- react & react-dom

They'll end up being bundled inside application bundle while others won't.

File: dist/index.html [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>Infozona</title>
+</head>
+<body>
+    <div id="app"></div>
+    <script src="./bundle.js"></script>
```
Comment (vladimyr) [inline dist/index.html]: Avoid filesystem-esque paths in html, ie change it to:
```html
   <script src="bundle.js"></script>
```

File: client/main.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+function Dates(dates) {
+  const dateList = dates.map(date => <Date date={date} />)
+  return <div className="dateList container">
+    <div>{ dateList }</div>;
+  </div>
+}
+
+Http.get('/api/calendar')
+  .then((response) => response.data)
+  .then((data) => Dates(data))
```
Comment (vladimyr) [inline client/main.js]: You can join those two steps:
```js
  .then(({ data }) => Dates(data))
```

File: client/main.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+import React from 'react';
+import { render } from 'react-dom';
+import Http from 'axios/dist/axios';
```
Comment (vladimyr) [inline client/main.js]: :bowtie: I understand where did this come from but it is simply wrong...
According to [source](https://github.com/mzabriskie/axios/blob/master/dist/axios.js#L94) axios exports prebaked axios instance that has `#create` factory attached alongside with `Axios` class exposed same way.
Pascal casing is used to indicate that value is a class while in reality you imported instance of a class 😵 
It would be really crazy to see something like:
```js
let client = new Http.Axios(options);
```
Long story short it should be `http`.

Additionally don't go into webpack's way. Bundling already bundled code disables any possible optimisation that webpack might apply to it. Webpack should warn you about that one. 🤔  Correct way is to include source variant ie:
```js
import http from 'axios'
```

**Review by vladimyr — CHANGES_REQUESTED** [review]
@kronicker I don't care much about underlaying server implementation but in this particular case I don't see any benefits migrating to express. On contrary it is just bigger and slower, ofc offers more but we are not using it that much?

**Comment by DorianSyruss** [comment]
Oh, I'm loving this ... 😆 😆 🛑

File: client/main.js (hunk 10 above)
Comment (kronicker) [reply client/main.js]: I agree this was following some unnecessary (or even wrong) conventions.
About the route, `axios` imports of node builtin modules was confusing `Rollup` which was used in the first run and this was the only way around it. But it should've been changed after `webpack` switch! Mea culpa 😄

File: client/main.js (hunk 9 above)
Comment (kronicker) [reply client/main.js]: 🏆

File: dist/index.html (hunk 8 above)
Comment (kronicker) [reply dist/index.html]: This is dynamically added by `webpack-html-template`, that is the usual way that the plugin injects script tags. Problem here is that it is pushed to origin in the first place and I'm not sure why since it `dist` folder is added to `.gitignore`.

File: client/main.js (hunk 10 above)
Comment (kronicker) [reply client/main.js]: Also since I'm used to using airbnb style guide I failed to find a recommendation for this example. The closest one I found is [this](https://github.com/airbnb/javascript#naming--PascalCase-singleton), am I correct?

File: package.json (hunk 7 above)
Comment (kronicker) [reply package.json:50]: Wasn't sure myself since I understood you that only dependencies which are needed to run the app in production need to be in dependencies and since those are bundled during build time they are not needed to run the app?

Comment (vladimyr) [reply package.json:50]: That's true if we deploy prebuilt bundle to prod which is rarely the case. But then you need rest of dev dependencies to bundle stuff 🤣
My own personal code is following. It is technically irrelevant how you spread dependencies because most of the time you want to install both. Because there isn't any tech "law" there I decided to create semantic one and use it to communicate to other devs:
- if it is in deps it means it is required from code (server or client)
- if it is in development deps it means it isn't (ie might be required from webpack or for example from seed script or something like that)

File: dist/index.html (hunk 8 above)
Comment (vladimyr) [reply dist/index.html]: Well I guess it doesn't matter then. Did you try with `git rm --cached dist/index.html`?

File: client/main.js (hunk 10 above)
Comment (vladimyr) [reply client/main.js]: >23.3 Use PascalCase only when naming constructors or classes. eslint: new-cap jscs: requireCapitalizedConstructors

From airbnb style guide 😉

**Comment by vladimyr** [comment]
@DorianSyruss https://youtu.be/Tn_fdnnUEcI 😁

File: client/main.js (hunk 10 above)
Comment (kronicker) [reply client/main.js]: Thx 👍

File: webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    devtool: 'eval',
+    devServer: {
+      contentBase: path.join(__dirname, "dist"),
+      noInfo: true,
+      proxy: {
+        '/api': `http://localhost:${port}`
+      }
+    }
+  };
+  return Object.assign(dev, baseConfig);
```
Comment (vladimyr) [inline webpack.config.js]: I believe you meant:
```js
Object.assign({}, baseConfig, dev);
```
The way you did it values from base might override dev. Luckily that won't happen due to non-existent property override in this particular case but still 👔

**Comment by kronicker** [comment]
@vladimyr I find these reviews super helpful and joyous, no misery on my behalf 💯 😂 

On a more on-topic note, I chose express since the further plan was to set up a database and a sync job so that scraping is done periodically by the server and data is inserted to db rather than waiting for scraper every time a user loads the page. Also this way we can build our own archive. 
Keeping all that in mind I figured it would be easier to do it with express and to migrate to it at this point in time while the code base is still rather simple. 
If you don't agree with the vision or the tooling I'll be more than happy to revert the change.

p.s: My apologies for some of the sloppiness in the code, didn't manage to do a proper recap before requesting PR.

File: webpack.config.js (hunk 11 above)
Comment (kronicker) [reply webpack.config.js]: It is exactly what I meant. Uptime is taking its toll 😛

**Comment by vladimyr** [comment]
@kronicker Makes perfect sense. I didn't have any vision but that's​ really​ awesome one 👌
I'm glad that I've managed to bring some fun into this 🙃

File: package.json [hunk 12]
```diff
@@ -2,19 +2,51 @@
...
+  "devDependencies": {
+    "babel-core": "^6.24.1",
+    "babel-loader": "^7.0.0",
+    "babel-plugin-external-helpers": "^6.22.0",
+    "babel-preset-es2015": "^6.24.1",
+    "babel-preset-react": "^6.24.1",
+    "bootstrap-loader": "^2.1.0",
+    "bootstrap-sass": "^3.3.7",
+    "css-loader": "^0.28.1",
+    "date-fns": "^1.28.4",
```
Comment (vladimyr) [inline package.json]: Missed this one? 🙈

Comment (kronicker) [reply package.json]: Holly 💩, I'm annoying! Won't be doing PRs this late no more 🙈 💤

**Comment by vladimyr** [comment]
@kronicker 🏅🎉🏄

**Comment by DorianSyruss** [comment]
I don't agree with the moral aspect of storing those data in db... say what you want 😢

**Comment by vladimyr** [comment]
reaction
